atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TokenRangeWritesLogger.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableRangeMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.util.MathUtils;
+
+public final class TokenRangeWritesLogger {
+    private static final Logger log = LoggerFactory.getLogger(TokenRangeWritesLogger.class);
+
+    @VisibleForTesting
+    static final long THRESHOLD_WRITES_PER_TABLE = 1_000_000L;
+    @VisibleForTesting
+    static final long TIME_UNTIL_LOG_MILLIS = 24 * 60 * 60 * 1000L;
+    private static final double CONFIDENCE_FOR_LOGGING = 0.99;
+
+    private final ConcurrentMap<TableReference, TokenRangeWrites> statsPerTable = new ConcurrentHashMap<>();
+    private volatile Set<Range<LightweightOppToken>> ranges = ImmutableSet.of(Range.all());
+
+    private TokenRangeWritesLogger() {
+        // use factory method
+    }
+
+    public static TokenRangeWritesLogger createUninitialized() {
+        return new TokenRangeWritesLogger();
+    }
+
+    public void updateTokenRanges(Set<Range<LightweightOppToken>> newRanges) {
+        Preconditions.checkArgument(!newRanges.isEmpty(), "Set of ranges must not be empty!");
+        if (!Sets.symmetricDifference(ranges, newRanges).isEmpty()) {
+            ranges = newRanges;
+            statsPerTable.clear();
+        }
+    }
+
+    <V> void  markWritesForTable(Set<Map.Entry<Cell, V>> entries, TableReference tableRef) {
```

Comment:
nit: double space?

Reply:
Haven't seen where this is called from but could we just pass `Map<Cell, V>`?

Reply:
This (or using a multiset, but this was simpler) used to be necessary in a previous version, because there could be duplicate `Cell`s. Now simplified!